nomad/resource_job.go

```diff
@@ -169,27 +195,133 @@ func resourceJobRegister(d *schema.ResourceData, meta interface{}) error {
 		return fmt.Errorf("error applying jobspec: %s", err)
 	}
 
+	log.Printf("[DEBUG] job '%s' registered", *job.ID)
 	d.SetId(*job.ID)
 	d.Set("name", job.ID)
 	d.Set("modify_index", strconv.FormatUint(resp.JobModifyIndex, 10))
 
+	evalId := resp.EvalID
+	if !d.Get("detach").(bool) {
+		log.Printf("[DEBUG] will monitor deployment of job '%s'", *job.ID)
+		stateConf := &resource.StateChangeConf{
+			Pending:    []string{"monitoring_deployment", "monitoring_evaluation"},
+			Target:     []string{"job_scheduled_without_deployment", "deployment_successful"},
+			Refresh:    deploymentStateRefreshFunc(client, evalId),
+			Timeout:    timeout,
+			Delay:      10 * time.Second,
+			MinTimeout: 3 * time.Second,
+		}
+
+		state, err := stateConf.WaitForState()
+		if err != nil {
+			return fmt.Errorf(
+				"error waiting for job '%s' to deploy successfully: %s",
+				*job.ID, err)
+		}
+
+		deployment := state.(*api.Deployment)
+
+		if deployment != nil {
+			d.Set("deployment_id", deployment.ID)
+			d.Set("deployment_status", deployment.Status)
+		} else {
+			d.Set("deployment_id", "")
+			d.Set("deployment_status", "")
+		}
+	}
+
 	return resourceJobRead(d, meta) // populate other computed attributes
 }
 
+type monitorState struct {
+	// evalId is the evaluation that we are currently monitoring. This will change
+	// along with follow-up evals.
+	evalId string
+
+	// deploymentId is the deployment that we are monitoring. This is captured from the
+	// final evaluation.
+	deploymentId string
+}
+
+// deploymentStateRefreshFunc returns a resource.StateRefreshFunc that is used to watch
+// the deployment from a job create/update
+func deploymentStateRefreshFunc(client *api.Client, initialEvalId string) resource.StateRefreshFunc {
+	watching := &monitorState{
```

Comment:
The `monitorState` struct seems only used here.  What's its benefit compared to having `evalId` and `deploymentId` variables here without the struct?

Reply:
i had originally (and incorrectly) thought i could have terraform pass it in, so i had it as a struct. but i was wrong, and I moved it into this function factory. 

this can be exploded it into multiple variables, but the risk there is that someone adding a new tracked data must remember to make that data a pointer so that it persists across calls to the refresh function, whereas I've taken care of that.